content/languages/cpp/igloo/stringizers.md

```diff
@@ -8,3 +8,438 @@ tags:
+    Actual: [ [unsupported type], [unsupported type] ]
+```
+
+Not only your custom types can be affected by this issue. Many built-in, standard, or 3rd party types also cannot be stringified. Basically, every type which does not define its version of output stream operator (`operator <<`) is affected, and `std::pair` template is a very common case of such type for Codewars kata.
+
+## Solutions
+
+To make a stringification of unsupported types possible, you have to provide one (or both) of code snippets: definition of `operator <<` for your type, or specialization of `snowhouse::Stringizer<T>` template. Snippets should be located just where your custom types are, usually in **Preloaded** part.
+
+### Stringification with `operator <<`
+
+`operator <<` is the easiest option to provide stringification, but it can be used only with types you control. You can add `operator <<` for types you created for the kata, but not for 3rd party, external, or standard types, like for example `std::pair`. For them, you have to use specialized `Stringizer<T>` ([see below](#stringification-with-stringizert)).
+
+Definition of `operator <<` for Snowhouse does not differ from implementation of any other output stream operator for C++ types:
+
+```cpp
+//Preloaded section
+
+//custom type
+struct Point2d {
+  int x, y;
+  bool operator== (const Point2d& other) const {
+    return x == other.x && y == other.y;
+  }
+};
+
+//output stream operator
+std::ostream& operator<<(std::ostream& stream, const Point2d& a)
+{
+  stream << '(' << a.x << ", " << a.y << ')';
+  return stream;
+}
+```
+
+Such definition causes that your custom type is now displayed properly, even if it's an element of a container:
+
+```cpp
+//Tests
+
+It(should_pretty_print_custom_class_with_ostream_operator)
+{
+  Point2d start {0,  0};
+  Point2d end   {7, -2};
+  Assert::That(start, Equals(end));
+}
+
+It(should_pretty_print_vector_of_elements_of_custom_type_with_ostream_operator)
+{
+  std::vector<Point2d> actual   = { {0, 0}, {2, 2}  };
+  std::vector<Point2d> expected = { {1, 1}  };      
+  Assert::That(expected, Equals(actual));
+}
+```
+
+Now your custom types are displayed properly in assertion messages:
+
+```
+  should_pretty_print_custom_class_with_ostream_operator
+    Expected: equal to (7, -2)
+    Actual: (0, 0)
+  should_pretty_print_vector_of_elements_of_custom_type_with_ostream_operator
+    Expected: equal to [ (0, 0), (2, 2) ]
+    Actual: [ (1, 1) ]
+```
+
+
+### Stringification with `Stringizer<T>`
+
+Sometimes, definition of `operator <<` cannot be used for stringification. You cannot add it to types which you do not own, or you cannot re-define it types which already have it defined, but not in a way you'd like to. Or maybe you just prefer this way rather than `operator <<`. In such cases, Snowhouse framework allows you to use a `snowhouse::Stringizer<T>` template specialized for the type you want to stringify.
```

Comment:
```suggestion
Sometimes, definition of `operator <<` cannot be used for stringification. You cannot add it to types which you do not own, or you cannot re-define it for types which already have it defined, but not in a way you'd like to. Or maybe you just prefer this way rather than `operator <<`. In such cases, Snowhouse framework allows you to use a `snowhouse::Stringizer<T>` template specialized for the type you want to stringify.
```